### PR TITLE
vault contract added

### PR DESCRIPTION
**Pull request by akhilesh-singh-Lab49**
vault contract added

### CODE REVIEW

File: deployment/state/contracts.json [hunk 1]
```diff
@@ -169,5 +169,12 @@
...
         "transparentProxyId": "0.0.49057304"
+    },
+    {
+        "name": "governortransfertoken",
+        "id": "0.0.49089046",
+        "address": "0x0000000000000000000000000000000002ed0a16",
+        "timestamp": "2022-12-14T14:59:16.849Z",
+        "hash": "6655a441d909ea0bc16697dccd9d14c5"
     }
 ]
```
Comment (abhishek-k-sharma-lab49) [inline deployment/state/contracts.json:180]: This change does not belong to the PR, I think.

File: contracts/Incentive/Vault.sol [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        if((userContribution[msg.sender].lockTimeStart + lockPeriod) < block.timestamp) {
+            withdraw(_startPosition, _amount);
+            return (block.timestamp,userContribution[msg.sender].lockTimeStart,lockPeriod);
+        } else {
+            revert("you can't unlock your token because the lock period is not reached");
+        }
+    }
+
+    function claimAllReward(uint _startPosition) public returns (uint, uint) { //claim
+        for(uint i = _startPosition; i < tokenAddress.length && i < _startPosition + 10; i++){
```
Comment (shivamgarglab49) [inline contracts/Incentive/Vault.sol:126]: any reason to adding 10 here .. (_startPosition + 10)

File: contracts/Incentive/IVault.sol [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8;
+
+interface IVault {
+
+    function addToken(address _token, uint _amount) external;
+    function getTVL() external view returns(uint);
```
Comment (amandeeprehal-lab49) [inline contracts/Incentive/IVault.sol:8]: Pls rename to getTotalVolume

File: contracts/Incentive/Vault.sol [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+import "../common/IERC20.sol";
+import "../common/hedera/HederaTokenService.sol";
+import "prb-math/contracts/PRBMathUD60x18.sol";
+import "../common/hedera/HederaResponseCodes.sol";
+
+contract Vault is HederaResponseCodes {
+
+    using PRBMathUD60x18 for uint256;
+
+    IERC20 public stakingToken;
```
Comment (amandeeprehal-lab49) [inline contracts/Incentive/Vault.sol:15]: Pls make it private

File: contracts/Incentive/Vault.sol [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    uint lockPeriod;
+
+    uint public totalTokens;
```
Comment (amandeeprehal-lab49) [inline contracts/Incentive/Vault.sol:18]: Same here, no state should be public

File: contracts/Incentive/Vault.sol [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    address[] tokenAddress;
+    
+    address public owner;
+    IBaseHTS tokenService;
+
+    struct UserInfo {
+        uint num_shares;
```
Comment (amandeeprehal-lab49) [inline contracts/Incentive/Vault.sol:25]: Pls. change to numberShare or more meaning full. Pls don't use _

File: contracts/Incentive/Vault.sol [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+        mapping(address => uint) lastClaimedAmountT;
+        uint lockTimeStart;
+        bool exist;
+    }
+
+    struct RewardsInfo {
```
Comment (amandeeprehal-lab49) [inline contracts/Incentive/Vault.sol:31]: Pls ensure there shouldn't any un-used code. Do we have locking functionality?

File: contracts/Incentive/Vault.sol [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+        uint amount;
+        bool exist;
+    }
+
+    mapping(address =>  UserInfo) public userContribution;
```
Comment (amandeeprehal-lab49) [inline contracts/Incentive/Vault.sol:36]: Pls keep declaration at single block.  Let move struct definition to top.

File: test/vault-test.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    const MockBaseHTS = await ethers.getContractFactory("MockBaseHTS");
+    const mockBaseHTS = await MockBaseHTS.deploy(true, false);
+    mockBaseHTS.setFailType(0);
+
+    const TokenCont = await ethers.getContractFactory("ERC20Mock");
+    const tokenCont = await TokenCont.deploy(total, 0);
+
+    const signers = await ethers.getSigners();
+
+    // await tokenCont.setUserBalance(signers[0].address);
```
Comment (amandeeprehal-lab49) [inline test/vault-test.ts:21]: ?

File: contracts/incentives/Vault.sol [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        address _stakingToken,
+        uint256 _lockPeriod,
+        IBaseHTS _tokenService
+    ) public initializer {
+        require(
+            _stakingToken != address(0),
+            "Staking token should not be empty."
+        );
+        owner = msg.sender;
+        console.log(owner);
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:52]: Pls remove this console.

File: contracts/incentives/Vault.sol [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+pragma abicoder v2;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "../common/IBaseHTS.sol";
+import "../common/IERC20.sol";
+import "../common/hedera/HederaTokenService.sol";
+import "prb-math/contracts/PRBMathUD60x18.sol";
+import "../common/hedera/HederaResponseCodes.sol";
+
+import "hardhat/console.sol";
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:12]: Pls remove

File: contracts/incentives/Vault.sol [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    struct RewardsInfo {
+        uint256 amount;
+        bool exist;
+    }
+
+    using PRBMathUD60x18 for uint256;
+    IERC20 private stakingToken;
+    uint256 private lockPeriod;
+    uint256 private totalTokens;
+    address[] private tokenAddress;
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:31]: plural pls.

File: contracts/incentives/Vault.sol [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+                responseCode == HederaResponseCodes.SUCCESS,
+                "Vault: Add stake failed when contributor does not exist."
+            );
+            userContribution[msg.sender].shares = _amount;
+            userContribution[msg.sender].exist = true;
+            userContribution[msg.sender].lockTimeStart = block.timestamp;
+            totalTokens += _amount;
+            return block.timestamp;
+        } else {
+            claimAllReward(0);
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:88]: Why this call? `claimAllReward(0);`

File: contracts/incentives/Vault.sol [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+
+    struct RewardsInfo {
+        uint256 amount;
+        bool exist;
+    }
+
+    using PRBMathUD60x18 for uint256;
+    IERC20 private stakingToken;
+    uint256 private lockPeriod;
+    uint256 private totalTokens;
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:30]: Its stakingTokenTotalAmount, 

`totalTokens` doesn't tell anything.

File: contracts/incentives/Vault.sol (hunk 12 above)
Comment (amandeeprehal-lab49) [reply contracts/incentives/Vault.sol:31]: rewardTokens

File: contracts/incentives/Vault.sol [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+    address[] private tokenAddress;
+    address private owner;
+    IBaseHTS private tokenService;
+    mapping(address => UserInfo) public userContribution;
+    mapping(address => RewardsInfo) public rewardsAddress;
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:35]: rewards

File: contracts/incentives/Vault.sol [hunk 16]
```diff
@@ -0,0 +1,270 @@
...
+                int64(uint64(_amount))
+            );
+            require(
+                responseCode == HederaResponseCodes.SUCCESS,
+                "Vault: Add reward failed on token not exist."
+            );
+        }
+    }
+
+    function addToken(address _token, uint256 _amount) public {
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:139]: Pls remove this and make individual methods public.

File: contracts/incentives/Vault.sol [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+            "Withdraw failed."
+        );
+
+        userContribution[msg.sender].shares -= _amount;
+        totalTokens -= _amount;
+    }
+
+    function unlock(uint256 _startPosition, uint256 _amount)
+        internal
+        returns (
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:169]: Why do we need return type?

File: contracts/incentives/Vault.sol [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+        }
+    }
+
+    function claimAllReward(uint256 _startPosition)
+        public
+        returns (uint256, uint256)
+    {
+        for (
+            uint256 i = _startPosition;
+            i < tokenAddress.length && i < _startPosition + 10;
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:197]: Do we have batching library?

File: contracts/incentives/Vault.sol [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+        for (uint256 i; i < _token.length; i++) {
+            uint256 reward;
+            address token = _token[i];
+            if (userContribution[msg.sender].lastClaimedAmount[token] == 0) {
+                tokenService.associateTokenPublic(token, address(msg.sender));
+            }
+            reward = (rewardsAddress[token].amount -
+                userContribution[msg.sender].lastClaimedAmount[token]).mul(
+                    userContribution[msg.sender].shares
+                );
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:252]: Duplicate code.

File: contracts/incentives/Vault.sol [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+
+    function addReward(address _token, uint256 _amount) internal onlyOwner {
+        require(totalTokens != 0, "No token staked yet");
+        uint256 perShareRewards;
+        perShareRewards = _amount.div(totalTokens);
+        if (!rewardsAddress[_token].exist) {
+            tokenAddress.push(_token);
+            rewardsAddress[_token].exist = true;
+            rewardsAddress[_token].amount = perShareRewards;
+            tokenService.associateTokenPublic(_token, address(this));
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:113]: Duplicate code.

File: contracts/incentives/Vault.sol [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+        returns (uint256 timeStamp)
+    {
+        if (!userContribution[msg.sender].exist) {
+            for (uint256 i; i < tokenAddress.length; i++) {
+                address token = tokenAddress[i];
+                userContribution[msg.sender].lastClaimedAmount[
+                        token
+                    ] = rewardsAddress[token].amount;
+                tokenService.associateTokenPublic(token, address(msg.sender));
+            }
```
Comment (amandeeprehal-lab49) [inline contracts/incentives/Vault.sol:71]: Refactor exist and non-exist code

File: contracts/incentives/Vault.sol (hunk 20 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/incentives/Vault.sol:111]: done

File: contracts/incentives/Vault.sol (hunk 19 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/incentives/Vault.sol]: done

File: contracts/incentives/Vault.sol (hunk 17 above)
Comment (abhishek-k-sharma-lab49) [reply contracts/incentives/Vault.sol:158]: done

**Review by amandeeprehal-lab49 — APPROVED** [review]
